scheds/rust/scx_lavd/src/bpf/main.bpf.c

```diff
@@ -1761,6 +1816,44 @@ void BPF_STRUCT_OPS(lavd_dispatch, s32 cpu, struct task_struct *prev)
 	scx_bpf_consume(LAVD_GLOBAL_DSQ);
 }
 
+static u32 calc_cpuperf_target(struct sys_cpu_util *cutil_cur,
+			       struct task_ctx *taskc, struct cpu_ctx *cpuc)
+{
+	u64 max_load, cpu_load;
+	u32 cpuperf_target;
+
+	/*
+	 * If a task with an average performance criticality of the current CPU
+	 * consumes 100% of CPU time, we set the CPU performance to the
+	 * maximum.
+	 */
+	max_load = cutil_cur->avg_perf_cri * 1000 /* max cpu util */;
+	cpu_load = taskc->perf_cri * cpuc->util;
```

Comment:
If a CPU has a high average perf criticality, wouldn't that mean that we want the CPU to be running fast even if the current task doesn't happen to have high criticality? e.g. In a single CPU system, if there's an active task which would qualify for highest performance level, wouldn't that warrant keeping the clock high?

Reply:
Ah, it compares against the system-wide average performance criticality (`cutil_cur->avg_perf_cri`).
The key idea is as follows:

- We determine the clock frequency of a CPU using two factors: 1) the current CPU utilization (cpuc->util) and 2) the current task's performance criticality (taskc->perf_cri) compared to the system-wide average performance criticality (cutil_cur->avg_perf_cri).

- When a current CPU utilization is 100%, and the current task's performance criticality is the same as the system-wide average criticality, we set the target CPU frequency to the maximum.

- In other words, even if CPU utilization is not so high, the target CPU frequency could be high when the task's performance criticality is high enough (i.e., boosting CPU frequency). On the other hand, the target CPU frequency could be low even if CPU utilization is high when a non-performance-critical task is running (i.e., deboosting CPU frequency).

Reply:
For simplicity, let's assume a single CPU system. Let's say there are two tasks. One with low criticality and the other high. The low one is executing and the high one is right behind. The current logic would put the CPU in a lower clock state, right? Or is it that as the high perf criticality one is likely to have priority over the low one anyway, this mostly doesn't matter?

Reply:
Yes, your understanding is correct. If it happens, the clock frequency will go up in the next tick, so I assumed that doesn't matter much. One potential solution would be calling scx_bpf_cpuperf_set() at ops.running() too. My slight worry about such an approach is the cost of scx_bpf_cpuperf_set()—actually, the cost of scaling the driver's hook, especially when task's runtime is short (say 100 usec). If the cost of scx_bpf_cpuperf_set() is cheap enough, we can call scx_bpf_cpuperf_set() in two places: at ops.running() for immediate reaction and at ops.tick() for reflecting changed system load. What do you think?

Reply:
I don't have any practical experience with cpufreq, so my opinion doesn't really matter. I was just trying to read the code and imagine how it would work. In case they might be helpful, the followings are what I learned from reading kernel code and listening to someone who's experimenting with this internally:

- The schedutil and cpufreq code does have its own rate-limiting mechanism.
- But still, trying to switch the freq too often does have noticeable overhead.

The reason why I kept asking about the same question is because the in-kernel implementation seems to be 1. use per-cpu aggregate value to determine target freq 2. avoid ramping down freq if util is close to 100%. ie. it looked like one of the main goals of the in-kernel implementation is avoiding the total amount of work done compared to max-freq condition. I was just wondering whether there would be conditions where lavd's implementation would lose total amount of work performed. So, just a theoretical curiosity than anything else.

Reply:
Then I think it would be better to keep the code as it is. I also considered two things (from the book). Updating the target frequency at ops.tick() is a kind of rate-limiting with a bounded delay. I will come up with more sophisticate approach after reading the kernel code more to adopt a similar (or better) strategy.